drivers/serial/uart_ns16550.c

```diff
@@ -355,8 +362,12 @@ static int uart_ns16550_configure(struct device *dev,
 	}
 #endif
 
+	k_spin_unlock(&dev_data->lock, key);
```

Comment:
Instead of unlock->call->lock, you can rename `set_baud_rate()` to something like `set_baud_rate_unlocked()` which contains code without locking. Then add a "new" `set_baud_rate()` function which does locking, then call `set_baud_rate_unlocked()`, then unlock. So in here, `set_baud_rate_unlocked()` can be used, so there is no need to do unlock->call->lock.

Reply:
That makes sense. I'll work that into the update soon. Thanks!

Reply:
Can you please help me resolve this build error that I get with the above implementation? Since we would call `set_baud_rate_unlocked( )` here in `configure`, then the `set_baud_rate( )` would need to be called from an application to be considered used in code coverage. Please correct me if I'm wrong.
             warning: 'set_baud_rate' defined but not used [-Wunused-function]
      291 | static void set_baud_rate(struct device *dev, u32_t baud_rate)
FUNC_UNUSED was not recognized....

Reply:
I thought `set_baud_rate()` was a API function. However, since `set_baud_rate()` is only called with this file, there is no need to do spinlock within this function. Locking can be done in the caller, as it looks like those functions need to lock anyway.

Reply:
Thanks, @dcpleung! You're right, `set_baud_rate()` is an API function, so it should be spinlocked, yeah? The only place it was being called in this file was in `configure`, but this convo considers replacing that call with the `set_baud_rate_unlocked()`.

Reply:
also, sorry it's a build *warning* not error

Reply:
`set_baud_rate()` is only being called by `uart_ns16550_configure()` and `uart_ns16550_line_ctrl_set()`. It is not directly callable by app. So there is no need to lock inside `set_baud_rate()` as the other two callers need to lock anyway.

Reply:
Ok update here, I was mistaken about `set_baud_rate` is not an API function. It is not in uart.h, and it is also not assigned in the api structure in this driver `static const struct uart_driver_api uart_ns16550_driver_api`.
So, no need to spinlock within this function (as mentioned above) -- that means we do not need to  use `set_baud_rate_unlocked()`? Thanks for the help to figure this out!